### PR TITLE
MDEV-36641 Implement Oracle Compatibility - INITCAP function

### PR DESCRIPTION
**Pull request by bardiaHSZD**
INITCAP() implementation returns char, with the first letter of each word in uppercase, all other letters in lowercase. Words are delimited by white space or characters that are not alphanumeric.

The return value is the same data type as char. The database sets the case of the initial characters based on the binary mapping defined for the underlying character set.

Examples

The following example capitalizes each word in the string:

Input:
```
SELECT INITCAP('hello world')
```

Output:
```
Hello World
```
Summary of changes in files:

Code:

* sql/item_create.cc:
  * Register the INITCAP function
* sql/item_strfunc.cc:
  * Implement the INITCAP function
* sql/item_strfunc.h:
  * Define the INITCAP function

Unit Test(s):

* mysql-test/main/initcap_functionality.test
* mysql-test/main/initcap_functionality_utf.test

Copyright:

All new code of the whole pull request, including one or several files that are either
new files or modified ones, are contributed under the BSD-new license.
I am contributing on behalf of my employer Amazon Web Services, Inc.

### CODE REVIEW

File: mysql-test/main/func_initcap.test [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+#-- Test file: Special Character Tests
+SELECT INITCAP('hello123world') AS 'Expected "Hello123world"';
+SELECT INITCAP('123Hello') AS 'Expected "123Hello"';
+
+#-- Test file: NULL Handling Tests
+SELECT INITCAP(NULL) AS 'Expected NULL';
+
+#-- Test file: Edge Cases Tests
+SELECT INITCAP('') AS 'Expected ""';
```
Comment (svoj) [inline mysql-test/main/func_initcap.test:22]: Let's follow common practice of testing similar functions and move this test either to `func_str.test` or rename to `func_initcap.test`.

File: mysql-test/main/initcap_functionality.test [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#-- File: initcap_basic_functionality_tests.mtr
+
+#-- Test file: Basic Functionality Tests
```
Comment (svoj) [inline mysql-test/main/initcap_functionality.test]: Was it supposed to be `--echo # Basic Functionality Tests`?

File: mysql-test/main/initcap_functionality_utf.test [hunk 3]
```diff
@@ -0,0 +1,117 @@
+--source include/have_utf8mb4.inc
+
+--echo # Character Set Tests Setup
+SET NAMES utf8mb4;
+SET character_set_client = utf8mb4;
+SET character_set_connection = utf8mb4;
+SET character_set_results = utf8mb4;
```
Comment (svoj) [inline mysql-test/main/initcap_functionality_utf.test]: This section looks redundant. Any reason to have it?

File: mysql-test/main/initcap_functionality_utf.test [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+SET NAMES utf8;
+SELECT INITCAP('hello world') AS 'UTF-8 Basic Test';
+
+--echo # Testing with UTF8MB4
+SET NAMES utf8mb4;
+SELECT INITCAP('hello world 👋') AS 'UTF8MB4 with Emoji';
+SELECT INITCAP('中文测试') AS 'UTF8MB4 Chinese';
+
+--echo # Testing character set conversion
+SET NAMES utf8mb4;
```
Comment (svoj) [inline mysql-test/main/initcap_functionality_utf.test]: Redundant.

File: mysql-test/main/initcap_functionality_utf.test [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+SELECT INITCAP(CONVERT('hello world' USING latin1)) AS 'Converted Latin1';
+SELECT INITCAP(CONVERT('française' USING utf8)) AS 'Converted UTF8';
+
+--echo # Reset to default character set
+SET NAMES utf8mb4;
```
Comment (svoj) [inline mysql-test/main/initcap_functionality_utf.test]: Redundant.

File: mysql-test/main/initcap_functionality_utf.test [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+--echo # Special Character Tests
+SELECT INITCAP('hello123world') AS 'Expected "Hello123world"';
+SELECT INITCAP('123hello') AS 'Expected "123Hello"';
+SELECT INITCAP('hello世界123') AS 'Mixed with Numbers';
+
+--echo # NULL and Empty String Tests
+SELECT INITCAP(NULL) AS 'Expected NULL';
+SELECT INITCAP('') AS 'Expected ""';
+
+# Extended Unicode Tests
```
Comment (svoj) [inline mysql-test/main/initcap_functionality_utf.test]: No more `--echo #` like for the above comments?

File: mysql-test/main/initcap_functionality_utf.test [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+SELECT INITCAP('zdravo svet') AS 'Serbian Test - Expected: Zdravo Svet';
+SELECT INITCAP('ahoj svet') AS 'Slovak Test - Expected: Ahoj Svet';
+SELECT INITCAP('zdravo svet') AS 'Slovenian Test - Expected: Zdravo Svet';
+SELECT INITCAP('sveiki pasaule') AS 'Latvian Test - Expected: Sveiki Pasaule';
+SELECT INITCAP('labas pasauli') AS 'Lithuanian Test - Expected: Labas Pasauli';
+SELECT INITCAP('tere maailm') AS 'Estonian Test - Expected: Tere Maailm';
+SELECT INITCAP('вітаю свет') AS 'Belarusian Test - Expected: Вітаю Свет';
+SELECT INITCAP('გამარჯობა მსოფლიო') AS 'Georgian Test - Expected: გამარჯობა მსოფლიო';
+SELECT INITCAP('բարեւ աշխարհ') AS 'Armenian Test - Expected: Բարեւ Աշխարհ';
+SELECT INITCAP('salam dünya') AS 'Azerbaijani Test - Expected: Salam Dünya';
```
Comment (svoj) [inline mysql-test/main/initcap_functionality_utf.test]: Missing EOL, but test in general looks very nice.

File: sql/item_strfunc.cc [hunk 8]
```diff
@@ -51,6 +51,10 @@ C_MODE_END
 #include <sql_repl.h>
 #include "sql_statistics.h"
 #include "strfunc.h"
+#include <wchar.h>
+#include <wctype.h>
+#include <unordered_map>
+#include "../strings/ctype-unidata.h"
```
Comment (svoj) [inline sql/item_strfunc.cc]: None of these includes must be needed, right?

File: sql/item_strfunc.cc [hunk 9]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
 }
 
 
+bool Item_func_initcap::fix_length_and_dec(THD *thd)
+{
+  uint32_t char_length= args[0]->max_char_length();
+  if (agg_arg_charsets_for_string_result(collation, args, 1))
+    return TRUE;
+  DBUG_ASSERT(collation.collation != NULL);
+  fix_char_length(char_length);
```
Comment (svoj) [inline sql/item_strfunc.cc]: Feels like it should be
```
  multiply= collation.collation->casedn_multiply();
  fix_char_length_ulonglong((ulonglong) args[0]->max_char_length() * multiply);

```

We definitely should consider `multiply`, but we need feedback from other developers concerning `fix_char_length_ulonglong()`. It seems to be used for functions that may return large strings like `Item_func_ucase`.

File: sql/item_strfunc.cc [hunk 10]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
+  String *res= args[0]->val_str(&tmp_value);
+  CHARSET_INFO *cs= collation.collation;
+  my_wc_t wc;
+  int rc;
+  bool new_word= true;
+
+  if ((null_value= args[0]->null_value))
+    return 0;
+
+  if (str->alloc(MY_MAX(res->length(), 4 * cs->mbminlen)))
```
Comment (svoj) [inline sql/item_strfunc.cc]: Hmm... should be `res->length() * multiply`?

File: sql/item_strfunc.cc [hunk 11]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
 
 
+bool Item_func_initcap::fix_length_and_dec(THD *thd)
+{
+  uint32_t char_length= args[0]->max_char_length();
+  if (agg_arg_charsets_for_string_result(collation, args, 1))
+    return TRUE;
+  DBUG_ASSERT(collation.collation != NULL);
+  fix_char_length(char_length);
+
```
Comment (svoj) [inline sql/item_strfunc.cc]: Trailing spaces on empty line.

File: sql/item_strfunc.cc [hunk 12]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
+  if ((null_value= args[0]->null_value))
+    return 0;
+
+  if (str->alloc(MY_MAX(res->length(), 4 * cs->mbminlen)))
+    return &tmp_value;
+  str->set_charset(collation.collation);
+  char *to= (char *) str->ptr();
+  char *to_end= to + str->alloced_length();
+  char *from= (char *) res->ptr(), *end= from + res->length();
+
```
Comment (svoj) [inline sql/item_strfunc.cc]: Trailing spaces.

File: sql/item_strfunc.cc [hunk 13]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
+  if (str->alloc(MY_MAX(res->length(), 4 * cs->mbminlen)))
+    return &tmp_value;
+  str->set_charset(collation.collation);
+  char *to= (char *) str->ptr();
+  char *to_end= to + str->alloced_length();
+  char *from= (char *) res->ptr(), *end= from + res->length();
+  
+  /* Use the charset's casefold information directly */
+  MY_CASEFOLD_INFO *casefold= cs->casefold ? cs->casefold : &my_casefold_default;
+
```
Comment (svoj) [inline sql/item_strfunc.cc]: Trailing spaces.

File: sql/item_strfunc.cc [hunk 14]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
+      if (my_isalpha(cs, *from))
+      {
+        /* Convert single byte to wide char for processing */
+        my_wc_t wc_char= (my_wc_t)*from;
+        if (new_word)
+          my_toupper_unicode(casefold, &wc_char);
+        else
+          my_tolower_unicode(casefold, &wc_char);
+        my_wc_t new_wc= wc_char;
+
```
Comment (svoj) [inline sql/item_strfunc.cc]: Trailing spaces.

File: sql/item_strfunc.cc [hunk 15]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
+      /* Multibyte character */
+      from+= rc;
+      if (my_uni_isalpha(wc))
+      {
+        my_wc_t new_wc= wc;
+        if (new_word)
+          my_toupper_unicode(casefold, &new_wc);
+        else
+          my_tolower_unicode(casefold, &new_wc);
+
```
Comment (svoj) [inline sql/item_strfunc.cc]: Trailing spaces.

File: sql/item_strfunc.cc [hunk 16]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
+      {
+        my_wc_t new_wc= wc;
+        if (new_word)
+          my_toupper_unicode(casefold, &new_wc);
+        else
+          my_tolower_unicode(casefold, &new_wc);
+        
+        if ((rc= cs->wc_mb(new_wc, (uchar*) to, (uchar*) to_end)) <= 0)
+        {
+          /* If conversion fails, try original character */
```
Comment (svoj) [inline sql/item_strfunc.cc]: Why? If conversion fails, we should return an error, don't we?

File: sql/item_strfunc.cc [hunk 17]
```diff
@@ -3097,6 +3101,128 @@ String *Item_func_soundex::val_str(String *str)
...
+  
+  while (from < end)
+  {
+    if ((rc= cs->mb_wc(&wc, (uchar*) from, (uchar*) end)) <= 0)
+    {
+      if (to < to_end)
+        *to++= *from;
+      from++;
+      continue;
+    }
```
Comment (svoj) [inline sql/item_strfunc.cc:3142]: We need to check with other developers, but I believe we must have enough routines to avoid separate handling of single-byte and multi-byte character sets.

E.g. `my_ci_ctype()` will give you character length, type of character (lower/upper/numeric/other) for arbitrary character set. We also have `my_ci_caseup()` / `my_ci_casedn()`.

**Review by svoj — CHANGES_REQUESTED** [review]
There is a typo in commit message: "currentl". According to https://github.com/MariaDB/server/blob/main/CODING_STANDARDS.md lines in commit message must be under 72 characters.

File: sql/item_strfunc.cc (hunk 17 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc:3142]: Sure I Will implement using existing function.

File: sql/item_strfunc.cc (hunk 16 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: This section does not exist in the new PR.

File: sql/item_strfunc.cc (hunk 15 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: This will be fixed.

File: sql/item_strfunc.cc (hunk 14 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: This will be fixed.

File: sql/item_strfunc.cc (hunk 13 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: This will be fixed.

File: mysql-test/main/initcap_functionality_utf.test (hunk 5 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: This will be removed.

File: mysql-test/main/initcap_functionality_utf.test (hunk 4 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: This will be removed.

Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: This will be removed.

File: mysql-test/main/initcap_functionality_utf.test (hunk 3 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: You are right. This will be removed.

File: mysql-test/main/initcap_functionality.test (hunk 2 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality.test]: Absolutely.

File: mysql-test/main/func_initcap.test (hunk 1 above)
Comment (bardiaHSZD) [reply mysql-test/main/func_initcap.test:22]: Sure. It will be in the next PR.

Comment (abarkov) [reply mysql-test/main/func_initcap.test:22]: I propose func_initcat.test, as func_str.test is already overloaded.

File: sql/item_strfunc.cc (hunk 17 above)
Comment (abarkov) [reply sql/item_strfunc.cc:3142]: @svoj is correct. Please use functions provided by MariaDB character set library.

Instead of ``my_ci_ctype()``, you can use a C++ wrapper:

  int ctype(int *to, const uchar *s, const uchar *e) const
  {
    return (cset->ctype)(this, to, s, e);
  }

Its return value has the same convention with ``my_wb_wc()``, i.e. a positive character length on success, or a non-positive number on error.

The type of the character returned in "to" is a bit-or of the following values:

```
#define _MY_U   01      /* Upper case */
#define _MY_L   02      /* Lower case */
#define _MY_NMR 04      /* Numeral (digit) */
#define _MY_SPC 010     /* Spacing character */
#define _MY_PNT 020     /* Punctuation */
#define _MY_CTR 040     /* Control character */
#define _MY_B   0100    /* Blank */
#define _MY_X   0200    /* heXadecimal digit */
```
Note, letters that do not have upper/lower conversion have _MY_U and _MY_L set at the same time.

File: mysql-test/main/initcap_functionality_utf.test (hunk 3 above)
Comment (abarkov) [reply mysql-test/main/initcap_functionality_utf.test]: Just ``SET NAMES utf8mb4;`` should be enough.

File: sql/item_strfunc.cc (hunk 9 above)
Comment (abarkov) [reply sql/item_strfunc.cc]: The way proposed by @svoj looks correct for me.
Also, the implementation could go directly to item_strfunc.h.

File: sql/item_strfunc.cc (hunk 10 above)
Comment (abarkov) [reply sql/item_strfunc.cc]: Some letters can grow in octet length during case conversion.
So it should be:
```
uint factor= MY_MAX(cs->caseup_multiply(), casedn_multiply());
str->alloc(res->char_length() * factor * cs->mbmaxlen);
```

However, to avoid a slow char_length() call, it could be simplified to:
```
uint factor= MY_MAX(cs->caseup_multiply(), casedn_multiply());
str->alloc(res->length() * factor);
```

File: mysql-test/main/func_initcap.test (hunk 1 above)
Comment (bardiaHSZD) [reply mysql-test/main/func_initcap.test:22]: The test and result files have been renamed to func_initcat.test and func_initcat.result, in the new PR.

File: mysql-test/main/initcap_functionality.test (hunk 2 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality.test]: The new PR contains the echo print message.

File: mysql-test/main/initcap_functionality_utf.test (hunk 3 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: Only kept "SET NAMES utf8mb4;" in the new PR.

File: mysql-test/main/initcap_functionality_utf.test (hunk 4 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: The redundant section of test was removed.

File: mysql-test/main/initcap_functionality_utf.test (hunk 5 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: The redundant section has been removed.

File: mysql-test/main/initcap_functionality_utf.test (hunk 6 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: Added "--echo #" for all sections in the new PR.

File: mysql-test/main/initcap_functionality_utf.test (hunk 7 above)
Comment (bardiaHSZD) [reply mysql-test/main/initcap_functionality_utf.test]: EOL is fixed in the new PR.

File: sql/item_strfunc.cc (hunk 8 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: The redundant includes were removed.

File: sql/item_strfunc.cc (hunk 9 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: The multiply implementation has been applied in the new PR.

File: sql/item_strfunc.cc (hunk 11 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: No trailing spaces in the new PR on this line.

File: sql/item_strfunc.cc (hunk 10 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: The fix has been applied in the new PR

File: sql/item_strfunc.cc (hunk 12 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: The trailing spaces have been removed.

File: sql/item_strfunc.cc (hunk 13 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: The trailing spaces have been removed.

File: sql/item_strfunc.cc (hunk 15 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: The trailing spaces were removed.

File: sql/item_strfunc.cc (hunk 17 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc:3142]: The implementation has been simplified, based on the above feedback.

File: sql/item_strfunc.cc (hunk 14 above)
Comment (bardiaHSZD) [reply sql/item_strfunc.cc]: The trailing spaced were removed.

**Comment by bardiaHSZD** [comment]
The MTR failures in clang-20 and continuous-integration are unrelated to changes in this PR.

**Comment by bardiaHSZD** [comment]
> There is a typo in commit message: "currentl". According to https://github.com/MariaDB/server/blob/main/CODING_STANDARDS.md lines in commit message must be under 72 characters.

The commit typo is corrected.

File: mysql-test/main/func_initcap_utf.result [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+SELECT INITCAP('test中文') AS 'Mixed Text';
+Mixed Text
+Test中文
+# Special Character Tests
+SELECT INITCAP('hello123world') AS 'Expected "Hello123world"';
+Expected "Hello123world"
+Hello123world
+SELECT INITCAP('123hello') AS 'Expected "123Hello"';
+Expected "123Hello"
+123Hello
```
Comment (abarkov) [inline mysql-test/main/func_initcap_utf.result:43]: Oracle returns '123hello' in this case

File: mysql-test/main/func_initcap_utf.result [hunk 19]
```diff
@@ -0,0 +1,197 @@
+# Character Set Tests Setup
+SET NAMES utf8mb4;
```
Comment (abarkov) [inline mysql-test/main/func_initcap_utf.result:2]: Can you please rename the file to func_initcap_utf8mb4.test ?

File: mysql-test/main/func_initcap.result [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+SELECT INITCAP('hello_world') AS 'Expected "Hello_World"';
+Expected "Hello_World"
+Hello_World
+# Special Character Tests
+SELECT INITCAP('hello123world') AS 'Expected "Hello123world"';
+Expected "Hello123world"
+Hello123world
+SELECT INITCAP('123Hello') AS 'Expected "123Hello"';
+Expected "123Hello"
+123Hello
```
Comment (abarkov) [inline mysql-test/main/func_initcap.result:30]: Oracle returns '123hello' here.
It seems a digit is considered to be a word start.

Can you please check all examples for Oracle compatibility?
We're adding an Oracle compatibility function, right?

**Review by abarkov — CHANGES_REQUESTED** [review]
Please fix incompatibilies and check all examples in Oracle. Thanks.

**Comment by gkodinov** [comment]
Closing due to inactivity: it's been more than a month since @abarkov provided review comments that are not being addressed. Please re-open when/if you intend to continue working on this.
